subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/testsuites/TestSuitesDependenciesIntegrationTest.groovy

```diff
@@ -0,0 +1,233 @@
+            }
+        }
+        """
+
+        expect:
+        succeeds 'checkConfiguration'
+    }
+
+    def 'user can add dependencies to the implementation, compileOnly and runtimeOnly configurations of a suite'() {
+        given:
+        buildFile << """
+        plugins {
+          id 'java'
+        }
+
+        ${mavenCentralRepository()}
+
+        dependencies {
+            // production code requires commons-lang3 at runtime, which will leak into tests' runtime classpaths
+            implementation 'org.apache.commons:commons-lang3:3.11'
+        }
+
+        testing {
+            suites {
+                test {
+                    dependencies {
+                        implementation 'com.google.guava:guava:30.1.1-jre'
+                        compileOnly 'javax.servlet:servlet-api:3.0-alpha-1'
+                        runtimeOnly 'mysql:mysql-connector-java:8.0.26'
+                    }
+                }
+                integTest(JvmTestSuite) {
+                    // intentionally setting lower versions of the same dependencies on the `test` suite to show that no conflict resolution should be taking place
+                    dependencies {
+                        implementation project(':')
+                        implementation 'com.google.guava:guava:29.0-jre'
+                        compileOnly  'javax.servlet:servlet-api:2.5'
+                        runtimeOnly 'mysql:mysql-connector-java:6.0.6'
+                    }
+                }
+            }
+        }
+
+        tasks.named('check') {
+            dependsOn testing.suites.integTest
+        }
+
+        tasks.register('checkConfiguration') {
+            dependsOn test, integTest
+            doLast {
+                def testCompileClasspathFileNames = configurations.testCompileClasspath.files*.name
+                def testRuntimeClasspathFileNames = configurations.testRuntimeClasspath.files*.name
+
+                assert testCompileClasspathFileNames.containsAll('commons-lang3-3.11.jar', 'servlet-api-3.0-alpha-1.jar', 'guava-30.1.1-jre.jar')
+                assert !testCompileClasspathFileNames.contains('mysql-connector-java-8.0.26.jar'): 'runtimeOnly dependency'
+                assert testRuntimeClasspathFileNames.containsAll('commons-lang3-3.11.jar', 'guava-30.1.1-jre.jar', 'mysql-connector-java-8.0.26.jar')
+                assert !testRuntimeClasspathFileNames.contains('servlet-api-3.0-alpha-1.jar'): 'compileOnly dependency'
+
+                def integTestCompileClasspathFileNames = configurations.integTestCompileClasspath.files*.name
+                def integTestRuntimeClasspathFileNames = configurations.integTestRuntimeClasspath.files*.name
+
+                assert integTestCompileClasspathFileNames.containsAll(/*'commons-lang3-3.11.jar', */'servlet-api-2.5.jar', 'guava-29.0-jre.jar') // TODO possible bug; why is commons-lang3 present in testCompileClasspathFileNames but not integTestCompileClasspathFileNames?
```

Comment:
@tresat @big-guy a more obvious example of inconsistent resolution of the dependency on the root project.

Reply:
Resolved by d5a439e8359